### PR TITLE
BUG-69053

### PR DESCRIPTION
**Pull request by mridulm**
a) Dont use case class for mutable state.
b) Remove broadcast when not required.
c) Fix MT issues in connection cache.
d) Fix encapsulation for some of the classes.

+CC @weiqingy, please review, thanks

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseConnectionCache.scala [hunk 1]
```diff
@@ -30,73 +33,91 @@ import org.apache.spark.Logging
...
-    try {
-      connectionMap.synchronized {
-        if(closed)
-          return
-        closed = true
+    connectionMap.synchronized {
+      if (closed.get()) return
+      try {
         housekeepingThread.interrupt()
+        resetCache()
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseConnectionCache.scala]: This change means that after close(), getStat will return (0, 0, 0)
If that is not the expectation, I can change it to only close the connections and leave the stats alone.

Comment (tedyu) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseConnectionCache.scala]: Please leave the stats alone.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseConnectionCache.scala [hunk 2]
```diff
@@ -30,73 +33,91 @@ import org.apache.spark.Logging
...
+  // 'duration' specified
+  private[hbase] def sleep(duration: Long, allowInterrupt: Boolean = false, allowClosed: Boolean = false): Unit = {
+    val startTime = System.currentTimeMillis()
+    var remaining = duration
+    while (remaining > 0) {
+      try {
+        Thread.sleep(remaining)
+      } catch {
+        case ex: InterruptedException if allowInterrupt => return
+        case ex: Exception => // ignore
```
Comment (tedyu) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseConnectionCache.scala]: Why ignore the exception ?

**Comment by weiqingy** [comment]
Hi, @mridulm  a, b and d, LGTM. For "c", could you elaborate what MT issues you are fixing? The reason that current test cases (`testBasic`, `testWithPressureWithoutClose`, `testWithPressureWithClose`) do not use `HBaseConnectionCache.connectionMap.synchronized` is there will be only one thread when doing `assert`. Of course, it does not hurt to add `HBaseConnectionCache.connectionMap.synchronized`. I'm neutral.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseConnectionCache.scala (hunk 2 above)
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseConnectionCache.scala]: Exception is expected when there is an explicit interrupt.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseConnectionCache.scala (hunk 1 above)
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseConnectionCache.scala]: Will do

**Comment by mridulm** [comment]
@weiqingy The issues with (c) is not with the test suite specifically - though it exists there too.
Any shared variable which is accessed/modified from different threads needs to be protected : either synchronized or via locks (Atomic*) - timeout, closed, connectionMap : otherwise, updates need not necessarily be visible across threads.

File: core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseConnectionCache.scala (hunk 1 above)
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/sql/execution/datasources/hbase/HBaseConnectionCache.scala]: resetStats == false by default now, preserving the stats even after a close()

**Comment by bikassaha** [comment]
Should some/all of this go into branch-1 and brach-2.1

**Comment by mridulm** [comment]
@bikassaha Ah, did not know we were maintaining different versions.
Once this goes through, will raise PR's against the other branches too. Thanks for letting me know.

PS: weirdly, dont get notifications for these comments ... strange.

**Comment by weiqingy** [comment]
@mridulm Currently for SHC, except the master branch, we do not require people to raise PRs against other branches. If the PRs are important for other branches, we will do the backport. Of course, we do welcome people to raise PRs against all the branches. :) Thanks.

**Comment by mridulm** [comment]
Great, thanks for clarifying @weiqingy !

**Pull request merged**